### ISSUE TITLE
kops generated terraform generates multiple identical IDs in nodes_security_group

### ISSUE THREAD

**Issue by mbolek** [kubernetes/kops#10786]
**1. What `kops` version are you running? The command `kops version`, will display
 this information.**

Version 1.18.3 (git-11ec695516)

**2. What Kubernetes version are you running? `kubectl version` will print the
 version if a cluster is running or provide the Kubernetes version specified as
 a `kops` flag.**

Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50",

**3. What cloud provider are you using?**

AWS

**4. What commands did you run?  What is the simplest way to reproduce this issue?**

`kops update cluster --out=. --target=terraform`

Issue exists when you have multiple instance groups

**5. What happened after the commands executed?**

Generated terraform output code has multiple, identical IDs for `nodes_security_group_ids` i.e.:

```
locals {
node_security_group_ids      = [aws_security_group.nodes-dev2-k8s-local.id, aws_security_group.nodes-dev2-k8s-local.id, aws_security_group.nodes-dev2-k8s-local.id]
}
(...)
output "node_security_group_ids" {
  value = [aws_security_group.nodes-dev2-k8s-local.id, aws_security_group.nodes-dev2-k8s-local.id, aws_security_group.nodes-dev2-k8s-local.id]
}
```

**6. What did you expect to happen?**

Seems to me this came up after upgrading kops to the latest version. Previously I've used 1.17.x and the generated `kubernetes.tf` had a single value (even though we've already been using multiple instance-groups):

```
locals {
 node_security_group_ids      = [aws_security_group.nodes-dev2-k8s-local.id]
}
(...)
output "node_security_group_ids" {
  value = [aws_security_group.nodes-dev2-k8s-local.id]
}
```

**7. Please provide your cluster manifest. Execute
  `kops get --name my.example.com -o yaml` to display your cluster manifest.
  You may want to remove your cluster name and other sensitive information.**

```yaml
```

**8. Please run the commands with most verbose logging by adding the `-v 10` flag.
  Paste the logs into this report, or in a gist and provide the gist link here.**

**9. Anything else do we need to know?**

This is not a problem when you just want to create the cluster. In our case, however, we use outputs from kops in other modules to create i.e. aurora security group. In this case terraform uses output from kops' `node_security_group_ids` as in input for aurora module that is then used to create a SG with multiple entries, but ends up with trying to create the same resource multiple times (since the `node_security_group_ids` now contains the same ID for each and every instance-group we have):
```
  # aws_security_group_rule.allow-aurora-sg[2] will be created
  + resource "aws_security_group_rule" "allow-aurora-sg" {
      + from_port                = 3306
      + id                       = (known after apply)
      + protocol                 = "tcp"
      + security_group_id        = "sg-88bbe3c3"
      + self                     = false
      + source_security_group_id = "sg-01701f2fa7ee960c5"
      + to_port                  = 3306
      + type                     = "ingress"
    }

  # aws_security_group_rule.allow-aurora-sg[3] will be created
  + resource "aws_security_group_rule" "allow-aurora-sg" {
      + from_port                = 3306
      + id                       = (known after apply)
      + protocol                 = "tcp"
      + security_group_id        = "sg-88bbe3c3"
      + self                     = false
      + source_security_group_id = "sg-01701f2fa7ee960c5"
      + to_port                  = 3306
      + type                     = "ingress"
    }
```

I think the previous logic with returning a single value in `node_security_group_ids` was correct, since every instance-group (at least in our case) uses exactly the same SG and the current flow breaks our code.

**Comment by rifelpet** [comment]
/kind bug

The reported behavior is confirmed in our integration tests:

https://github.com/kubernetes/kops/blob/ec495195a7288d3b26550f62bb57eb77d0d6b0b5/tests/integration/update_cluster/launch_templates/kubernetes.tf#L29-L31

A workaround would be to wrap the output references in `toset()` which should deduplicate the values. We should probably be using sets or maps instead of lists anyways because ordering of this information is not relevant. 

I do think we should overhaul the terraform outputs we generate now that terraform can output complex objects too.

**Comment by mmerrill3** [comment]
I saw this too when I did my latest upgrades using kops 1.20.beta 1.  The subnets are duplicated as well.  I'll take a look.

**Comment by luarx** [comment]
This is a critical bug because I lost access to my cluster. After removing manually those duplicated values I could access again.

**Comment by denisgranha** [comment]
will this be fixed in a hotfix (1.18.x) or should we expect it only for >1.19 versions?

**Comment by olemarkus** [comment]
Unlikely we will do another release of 1.18, so expect this only for the versions this has already been merged into.
